feat: init pragma price feed ch (#831)
cairo-book/cairo-book: src/ch16-05-01-price-feeds.md

```

## Creating a Price Feed Contract

After adding the required dependencies for your project, you'll need to define a contract interface that includes the required pragma price feed entry point.

### Contract Interface

```rust,noplayground
{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_08_price_feed/src/lib.cairo:price_interface}}
```

Of the two public functions exposed in the `IPriceFeedExample`, the one necessary to interact with the pragma price feed oracle is the `get_asset_price` function, a view function that takes in the `asset_id` argument and returns a `u128` value.

### Pragma Dependency Import to Contract

```rust,noplayground
{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_08_price_feed/src/lib.cairo:pragma_lib}}
```

The snippet above shows the necessary imports you need to add to your contract module in order to interact with the Pragma oracle.

### Required Price Feed Function Impl in Contract

```rust,noplayground
{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_08_price_feed/src/lib.cairo:price_feed_impl}}
```

> The `get_asset_price` function is responsible for retrieving the price of the asset specified in the `asset_id` argument from Pragma Oracle. Inside the function, an instance of the `IPragmaDispatcher` is created which serves as a dispatcher for interacting with the Pragma Oracle. The `get_data_median` method is then called from the created instance by passing the `DataType::SpotEntry(asset_id)` as an argument and assigned to a variable named `output` of type `PragmaPricesResponse`. Finally, the function returns an `output.price` a `u128` which is the price of the `asset_id` requested.

Reviewer: ```suggestion
The `get_asset_price` function is responsible for retrieving the price of the asset specified by the `asset_id` argument from Pragma Oracle. The `get_data_median` method is called from the `IPragmaDispatcher` instance by passing the `DataType::SpotEntry(asset_id)` as an argument and its output is assigned to a variable named `output` of type `PragmaPricesResponse`. Finally, the function returns the price of the requested asset as a `u128`.
```

## Example Application Using Pragma Price Feed

```rust,noplayground
{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_08_price_feed/src/lib.cairo:here}}
```

> The code above is an example implementation of price feed using the Pragma oracle. The contract imports necessary modules and interfaces, including the `IPragmaABIDispatcher` for interacting with the Pragma oracle contract and the `ERC20ABIDispatcher` for interacting with the ETH ERC20 token contract.

Reviewer: ```suggestion
The code above is an example implementation of an applications consuming a price feed from the Pragma oracle. The contract imports necessary modules and interfaces, including the `IPragmaABIDispatcher` for interacting with the Pragma oracle contract and the `ERC20ABIDispatcher` for interacting with the ETH ERC20 token contract.
```

> The contract has a `const` that stores the token pair ID of `ETH/USD`, and a `Storage` struct that holds two fields `pragma_contract` and `product_price_in_usd`, then the constructor function initializes the `pragma_contract` address and sets the `product_price_in_usd` to 100.

Reviewer: ```suggestion
The contract has a `const` that stores the token pair ID of `ETH/USD`, and a `Storage` struct that holds two fields `pragma_contract` and `product_price_in_usd`. The constructor function initializes the `pragma_contract` address and sets the `product_price_in_usd` to 100.
```

### Pragma Dependency Import to Contract

```rust,noplayground
{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_08_price_feed/src/lib.cairo:pragma_lib}}
```

The snippet above shows the necessary imports you need to add to your contract module in order to interact with the Pragma oracle.

### Required Price Feed Function Impl in Contract

```rust,noplayground
{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_08_price_feed/src/lib.cairo:price_feed_impl}}
```

The `get_asset_price` function is responsible for retrieving the price of the asset specified by the `asset_id` argument from Pragma Oracle. The `get_data_median` method is called from the `IPragmaDispatcher` instance by passing the `DataType::SpotEntry(asset_id)` as an argument and its output is assigned to a variable named `output` of type `PragmaPricesResponse`. Finally, the function returns the price of the requested asset as a `u128`.

## Example Application Using Pragma Price Feed

```rust,noplayground
{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_08_price_feed/src/lib.cairo:here}}
```

> > Note: Pragma returns the value of different token pairs using the decimal factor of 6 or 8. You can convert the value to the required decimal factor by dividing the value by 10^decimal_factor.

Reviewer 2: You should format nicely 10^... 

Check the Data Type chapter md file you ll see how to use latex in md files

Author: done

# Price Feeds

> Price feeds enabled by an oracle serve as a bridge between real-world data feed and the blockchain. It provides real time pricing data that is aggregated from multiple trusted external sources ( e.g. crypto exchanges, financial data providers, etc. ) to the blockchain network.

Reviewer 2: i think "They provide" would be more consistent witht the first sentence which is plurial with "price feeds"

{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_08_price_feed/src/lib.cairo:price_feed_impl}}
```

> The `get_asset_price` function is responsible for retrieving the price of the asset specified in the `asset_id` argument from Pragma Oracle. Inside the function, an instance of the `IPragmaDispatcher` is created which serves as a dispatcher for interacting with the Pragma Oracle. The `get_data_median` method is then called from the created instance by passing the `DataType::SpotEntry(asset_id)` as an argument and assigned to a variable named `output` of type `PragmaPricesResponse`. Finally, the function returns a `u128` the price property of `output`, which is the requested asset price.

Reviewer 2: > Finally, the function returns a `u128` the price property of `output`, which is the requested asset price.

what does this mean?
